lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -44,6 +44,50 @@ sub _translate_cond($) {
     die "Unknown '$cond'";
 }
 
+sub _prepare_data_table {
+    my ($self, $n, $paths, $dir_rs, $needles_rs) = @_;
+    my $filename = $n->filename;
+    my $hash     = {
+        id        => $n->id,
+        directory => $n->directory->name,
+        filename  => $filename,
+    };
+    my $dir_path = $n->directory->path;
+    my $real_dir_id;
+    my $dir = $paths->{$dir_path};
+
+    if ($dir) {
+        $real_dir_id = $dir->{real_path_id} if $dir && $dir_path ne ($dir->{real_path} // '');
+    }
+    else {
+        my $real_path_id  = $n->directory->id;
+        my $dir_real_path = realpath($dir_path);
+        if ($dir_real_path && $dir_real_path ne $dir_path) {
+            my $real_dir = $dir_rs->find({path => $dir_real_path});
+            $real_dir_id = $real_path_id = $real_dir->id if $real_dir;
+        }
+        $dir = {
+            real_path    => $dir_real_path,
+            real_path_id => $real_path_id,
+        };
```

Comment:
this value should be stored in $paths and used for next search.

Reply:
I think we do this correctly because `$dir` is a reference to an element within `$paths`.

Reply:
if I did a test about this correctly, the `$dir` is an `undef` in this `else` block,  so the `$dir = {}` is not added into `$paths`

Reply:
Do you have an idea how to test this better? The automatic test is passed and also NYTProf shows the same amount of readlink calls.

Reply:
Do you have an idea how to test this better? The automatic test is passed and also NYTProf shows the same amount of readlink calls.

Reply:
I did the test in my local test environment. I added some log message in the `else` block and found the `$paths` was always empty. 
 In my test,  the picture1 shows that, if we use `$dir->{}`, the `realpath` will be called 25 times.  If we add the value to `$paths->{$dir_path} = {}` directly, the `realpath` will be called 1 time.  BTW, the needle page loads 25 needles at one time and all of them are belong to one needle directory which is a symlink.
![picture1](https://user-images.githubusercontent.com/11713397/76615747-97b7f200-655d-11ea-8446-aec491eefa19.png)
![picture2](https://user-images.githubusercontent.com/11713397/76615750-9981b580-655d-11ea-8131-25265472e2eb.png)

Reply:
I managed to come up to this story. Thanks for sharing your information. This helps as reference.

I gathered coverage from runs of t/ui/21-admin-needles.t and found 5 vs. 18 calls of reallink in case of my "broken" code and the fixed code.

I pushed the fix.